include/boost/geometry/algorithms/detail/buffer/piece_border.hpp

```diff
@@ -0,0 +1,509 @@
+        bool const result = is_convex(previous, current, m_ring->begin() + m_begin + 2, m_ring->begin() + m_end, strategy)
+            && is_convex(previous, current, m_originals.begin(), m_originals.begin() + m_original_size, strategy)
+            && is_convex(previous, current, offsetted_front, strategy)
+            && is_convex(previous, current, offsetted_second, strategy);
+
+        return result;
+    }
+
+    template <typename It, typename SideStrategy>
+    bool is_convex(Point& previous, Point& current, It begin, It end, SideStrategy const& strategy) const
+    {
+        for (It it = begin; it != end; ++it)
+        {
+            if (! is_convex(previous, current, *it, strategy))
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    template <typename SideStrategy>
+    bool is_convex(Point& previous, Point& current, Point const& next, SideStrategy const& strategy) const
+    {
+        typename SideStrategy::equals_point_point_strategy_type const
+            eq_pp_strategy = strategy.get_equals_point_point_strategy();
+
+        int const side = strategy.apply(previous, current, next);
+        if (side == 1)
+        {
+            // Next is on the left side of clockwise ring: piece is not convex
+            return false;
+        }
+        if (! equals::equals_point_point(current, next, eq_pp_strategy))
+        {
+            previous = current;
+            current = next;
+        }
+        return true;
+    }
+
+    template <int Direction>
+    inline void step_for_monotonicity(Point const& current, Point const& next)
+    {
+        if (geometry::get<Direction>(current) >= geometry::get<Direction>(next))
+        {
+            m_is_monotonic_increasing = false;
+        }
+        if (geometry::get<Direction>(current) <= geometry::get<Direction>(next))
+        {
+            m_is_monotonic_decreasing = false;
+        }
+    }
+
+    template <typename It>
+    void check_monotonicity(It begin, It end)
+    {
+        m_is_monotonic_increasing = true;
+        m_is_monotonic_decreasing = true;
+
+        if (begin == end || begin + 1 == end)
+        {
+            return;
+        }
+
+        It it = begin;
+        for (It previous = it++; it != end; ++previous, ++it)
+        {
+            step_for_monotonicity<0>(*previous, *it);
+        }
+    }
+
+    template <typename It>
+    inline void calculate_radii(Point const& center, It begin, It end)
+    {
+        typedef geometry::model::referring_segment<Point const> segment_type;
+
+        bool first = true;
+
+        // An offsetted point-buffer ring around a point is supposed to be closed,
+        // therefore walking from start to end is fine.
+        It it = begin;
+        for (It previous = it++; it != end; ++previous, ++it)
+        {
+            segment_type const s(*previous, *it);
+            radius_type const d = geometry::comparable_distance(center, s);
```

Comment:
This needs a strategy. But there is a problem with it because it cannot be derived from the intersection strategy since the intersection strategy is lower-level than the distance point/segment strategy I think, so there would be a cyclic-dependency. I know that because recently I looked into this while checking whether or not it is possible to use `buffer` with `cs_undefined` CS and it is not possible right now because internally, at the top-most level the default intersection strategy is created and internally the default `comparable_distance` is called. But anyway, the solution I came up with for this cyclic-dependency is to implement a "proper" CS-specific buffer strategy providing both CS-specific intersection and comparable_distance strategies. So this would be a greater change outside of the scope of this PR I think. Therefore I propose to leave this at is is for now and later I'll propose my solution for this issue (I'd like to propose some more changes too along with it). Do you agree?

Reply:
We discussed a while ago: I'm quite against adding sub-strategies to the `IntersectionStrategy`. There should be a proper `UbrellaStrategy` and this is *not* the IntersectionStrategy. Because why? It doesn't use most of the strategies defined there (I think it uses only `side` and nothing more). It's just the wrong place. Why should the intersection of two segments define a `point_in_polygon`? I just don't get that.

Well, I understand how it happened - there needed to be a place and this strategy was passed everywhere - so it was added there. But it is inconvenient and should be refactored.

So yes, if this could be fixed by moving all these sub-strategies to a proper umbrella strategy, that would be very welcome.

And that umbrella strategy can be passed to buffer indeed, and there will be no cyclic dependencies.

Reply:
A disclaimer: I agree that the `IntersectionStrategy` is not the place for all of the getters that currently are there. Below I describe the rationale behind the current state, not claiming that it should stay as it is because I think it should be changed in general. I'm not sure that the `UbrellaStrategy` is the best solution though.

> There should be a proper UbrellaStrategy and this is not the IntersectionStrategy

That would be a breaking change. Note that the `IntersectionStrategy` in this case is at the same time the "original" strategy and also the umbrella strategy containing all strategies needed in algorithms which the `IntersectionStrategy` is passed into since the inclusion of Boost.Geometry into Boost. So it is both kinds of strategy right now, simply because otherwise it'd break the existing code.

> Why should the intersection of two segments define a point_in_polygon?

As above. Because of the design of the library and the desire to not introduce breaking changes.

Consider e.g. set operations, let's say intersection of polygons. The strategy which was always required since the inclusion of Boost.Geometry was the `IntersectionStrategy`. In addition to this the library requires to pass one strategy for an algorithm and it also guarantees that the CS of the strategy "overwrites" the CS of the geometry. So in order to not break the users code we have to require passing the same strategy as usual, in this case the `IntersectionStrategy` and this strategy has to be the umbrella strategy as well because of all of the other strategies needed deeper in the algorithms.

It would be possible to ensure backward compatibility with "proper" `UmbrellaStrategy` though. We'd have to create the `UmbrellaStrategy` from another "original" strategy (e.g. umbrella strategy for set operations created from classically passed intersection strategy). Note however that this means that all strategies in the umbrella strategy would be created based on the original intersection strategy. And this means that the umbrella strategy would be not needed since the library would create all other strategies automatically anyway.

There is also an extensibility issue. What if we introduced a "proper" `UbrellaStrategy` and after a while another lower-level sub-strategy was needed internally in the algorithm? Would we require to pass another sub-strategy into the `UbrellaStrategy` changing the interface and therefore breaking user's code or would we create it automatically based on other strategies that were in the `UmbrellaStrategy` before the change? If the answer is the latter then I'd argue that the `UbrellaStrategy` is not needed in the first place since it should be possible to create any sub-strategy from any set of other sub-strategies.

Btw, I'd be ok with redesigning the strategies part to something different even if this would mean breaking changes. The `UbrellaStrategy` is the most straightforward thing to do. It would cause the problems I described above and probably other which we don't take into account right now. Right now, the `IntersectionStrategy` is partially the `UmbrellaStrategy` so we'd have similar problems with the "proper one". I'm not even sure if we'd solve the cyclic dependency across sub-strategies this way.

Note that all of the problems I mentioned are derived from the original design of the library so I'd consider to do something more radical like e.g. replacing the strategies mechanism with something else instead of adding another layer of strategies for the user to use. I'm talking about the interface here.

Of course I may be mistaken with all of the above so if you have a proposal we can create an Issue and talk about it.

Reply:
Thanks! I will reread this later and react on it, probably this weekend.

Reply:
@barendgehrels After giving it more thought I think we could indeed try with "proper" `UmbrellaStrategy` at least for one function and see how it works out to avoid over-engineering things. I think we could workaround the problems I mentioned above and the result would still be the code which is more clear than it is now (the strategies in particular which I agree should have well-defined purpose).

Reply:
I really welcome this idea!

Will we create an issue for it - to get it in our regular process?